src/Microsoft.OData.Edm/Validation/ValidationRules.cs

```diff
@@ -1611,7 +1611,7 @@ public static class ValidationRules
                               var principalType = principalProperties.ElementAtOrDefault(i).Type.Definition;
                               if (!(dependentType is BadType) && !(principalType is BadType) && !dependentType.IsEquivalentTo(principalType))
                               {
-                                  string errorMessage = Strings.EdmModel_Validator_Semantic_TypeMismatchRelationshipConstraint(navigationProperty.DependentProperties().ToList()[i].Name, navigationProperty.DeclaringEntityType().FullName(), principalProperties.ToList()[i].Name, principalEntityType.Name, "Fred");
+                                  string errorMessage = Strings.EdmModel_Validator_Semantic_TypeMismatchRelationshipConstraint(navigationProperty.DependentProperties().ToList()[i].Name, navigationProperty.DeclaringEntityType().FullName(), principalProperties.ToList()[i].Name, principalEntityType.Name);

```

Comment:
Just realized we're computing `principaleProperties.ToList()`  just to obtain the name. But on line 1611, we call `principalProperties.ElementAtOrDefault(i)`.  If `principalProperties.ElementAtOrDefault(i)` returns the same object as `principalProperties.ToLits()[i]` then we just probably re-use the object instead of computing `ToList()`:


```c#
var principalProperty = principalProperties.ElementAtOrDefault(i);
var principalType = principaleProperty.Type.Definition;

if (...)
{
   string errorMessage = ... principalProperty?.Name
}